Changing to Pyglet from Soloud for Sound

Cleptomania opened the pull request:
This change re-works the sound module a bit, primarily abstracting the back-end behind a common interface, and switching from the SoLoud library to Pyglet. It has been tested against Windows, Mac, and Linux with .wav, .mp3, and .ogg formats.

There are no API breaking changes and everything should be able to use it in the same way as previously. However there are some feature additions.

Added Features:

- Sound looping
- Checking if a sound is playing or not
- Retrieve the length of a sound(appeared to be broken with the SoLoud implementation)
- Get/Set the volume of a sound after starting playback(appeared to broken with the SoLoud implementation)

This should also greatly improve cross-platform support.

The ffmpeg binaries for Mac have been added in a similar fashion to how SoLoud's binaries were included. They can be seen in the "arcade/lib" folder. If ffmpeg is desired for Windows/Linux all one would need to do is drop the corresponding binaries into there and they would automatically be loaded. The Windows/Linux binaries have not been included due Pyglet having out of the box support for Windows/Linux

Cleptomania committed:
feat(sound): Initial Pyglet sound implementation

Cleptomania committed:
feat(sound): Initial Pyglet sound implementation

Cleptomania committed:
Update to automatically load ffmpeg

Cleptomania committed:
Update to automatically load ffmpeg

Cleptomania committed:
feat(sound): Final auto-loading of ffmpeg. Also added Windows ffmpeg binaries.

Cleptomania committed:
feat(sound): Final auto-loading of ffmpeg. Also added Windows ffmpeg binaries.

Cleptomania committed:
Moved "libs" into arcade module

Cleptomania committed:
Moved "libs" into arcade module

Cleptomania committed:
moved ffmpeg library loading to main __init__ so that it happens before Pyglet import

Cleptomania committed:
moved ffmpeg library loading to main __init__ so that it happens before Pyglet import

Cleptomania committed:
tests(sound): Increased test coverage for sound module

Cleptomania committed:
tests(sound): Increased test coverage for sound module

Cleptomania committed:
tests(sound): Improvements to duration checking tests

Cleptomania committed:
tests(sound): Improvements to duration checking tests

Cleptomania committed:
tests(sound): Improvements for cross-platform checking of sound durations

Cleptomania committed:
tests(sound): Improvements for cross-platform checking of sound durations

Cleptomania committed:
tests(sound): more duration checking fixes

Cleptomania committed:
tests(sound): more duration checking fixes

Cleptomania committed:
tests(sound): removed duration checking tests

Cleptomania committed:
tests(sound): removed duration checking tests

Cleptomania committed:
build: Pinned minimum Pyglet version of 1.5.2 to ensure full support for Windows/Linux without using ffmpeg

Cleptomania committed:
build: Removed ffmpeg binaries for Windows

Cleptomania committed:
build: Removed ffmpeg binaries for Windows

Cleptomania committed:
build: Added maximum Pyglet version to avoid updating to Pyglet 2

Cleptomania committed:
docs(sound): Improved comments for the Sound module

Cleptomania committed:
docs(sound): Improved comments for the Sound module

einarf commented:
We should probably try to get this one merged soon. There are some PyWeekers bumping into audio problems.

pvcraven committed:
Highlight imporant lines in move-by-keyboard example.

pvcraven committed:
Update sprite movement by keyboard example code.

pvcraven committed:
Update examples around moving animations, and facing left/right.

einarf committed:
Index buffer example: template with shader

einarf committed:
Bug: num_vertices incorrect when index buffer is used

einarf committed:
Geometry: Remove old error check

einarf committed:
Example: Working index buffer example

einarf committed:
Geometry: Support 8, 16 and 32 bit index buffers

einarf committed:
Index buffer example

einarf committed:
Tests: Missing BufferDescription checks

einarf committed:
Tests: 8, 16 and 32 bit index buffers

eruvanos committed:
Fix gui tests (missing 'on_resize' and 'on_update' event type on MockWindow) (#775)

einarf committed:
gl: Missing type Context type hints

einarf committed:
arcade.gl: Improve module description

autodoc will fetch the first line in the docstring. We should make that as precise as possible

einarf committed:
Spelling errors

einarf committed:
Remove redundant tobytes() calls

We support the buffer protocol..

einarf committed:
docs: gl module warning should be more visible

einarf committed:
docs: customize arcade.gl index

Cleptomania commented:
Worth noting if https://github.com/pythonarcade/arcade/pull/780 gets merged then the `MANIFEST.in` file will need updated to remove packaging the soloud binaries and add packaging ffmpeg binaries for mac.

einarf committed:
gl: Instancing example

einarf committed:
Update instancing.py

einarf committed:
Tessellation shader support

einarf committed:
Tessellation example: Bezier curve

einarf committed:
draw_commands: Specify return types in docs

einarf committed:
ArcadeContext: Allow loading tessellation shaders from file

einarf committed:
Window: Allow specifying opengl version

einarf committed:
Tessellation example should request OpenGL 4.1

einarf committed:
release notes spelling error

einarf committed:
Release notes mockup (unreleased)

Cleptomania committed:
Structural changes to better support rapidly playing sounds

Cleptomania committed:
Structural changes to better support rapidly playing sounds

einarf committed:
BufferDescription: Sanity check buffer parameter

SirGnip committed:
Fix incorrect PyInstaller data file path handling docs #773 (#774)

einarf committed:
Added Texture.byte_size + tests

einarf committed:
Merge branch 'development' of https://github.com/pythonarcade/arcade into development

akapkotel committed:
Issue #782 Using shapely to find line of sight (#783)

* issue #782 faster checking for line of sight between Sprites

* issue #782 faster checking for line of sight between Sprites

1. shapely and setuptools removed from requirements.txt

* Update requirements.txt

removed all versioning.

Co-authored-by: akapkotel <btcuserbtc@gmail.com>

akapkotel committed:
Merge remote-tracking branch 'upstream/development' into using_shapely_to_find_line_of_sight (#788)

1. Replaced all geometry collision test with shapely functions.
2. Replaced math.sqrt function with math.hypot in get_distance function because hypot is faster - see math module documentation.
3. Removed some redundant local assignments in sprite_list.py functions with straightforward returns.

Co-authored-by: akapkotel <btcuserbtc@gmail.com>

Cleptomania committed:
Update setuptools configuration to align with PEP 517/518 (#780)

* Updates to setuptools configuration to align with PEP 517

* chore(setup): remove trailing comma

* Pin crucial package versions

* Update MANIFEST to include .typed files.

* Finalize package version pins.

Cleptomania committed:
Merge branch 'sound-pyglet' of github.com:pythonarcade/arcade into sound-pyglet

Cleptomania committed:
Update build MANIFSET to remove SoLoud and add ffmpeg binaries
